File: pkg/java/src/test/java/dev/openfga/language/FgaModToJsonShould.java
package dev.openfga.language;

import static java.util.stream.Collectors.joining;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;
import static org.junit.jupiter.params.provider.Arguments.arguments;

import dev.openfga.language.errors.ModFileValidationError;
import dev.openfga.language.errors.ModFileValidationSingleError;
import dev.openfga.language.util.TestsData;
import java.util.List;
import java.util.stream.Stream;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class FgaModToJsonShould {

    @ParameterizedTest(name = "{0}")
    @MethodSource("fgaModTestCases")
    public void transformFgaMod(
            String name, String fgaMod, String json, boolean skip, List<ModFileValidationSingleError> expectedErrors)
            throws Exception, ModFileValidationError {
        Assumptions.assumeFalse(skip);

        if (json != null) {
            var generatedJson = FgaModTransformer.transform(fgaMod);

            var expected = JSON.stringify(JSON.parse(json, FgaModFile.class));

            assertThat(generatedJson).isEqualTo(expected);
        } else {
            var thrown = catchThrowable(() -> FgaModTransformer.transform(fgaMod));

            assertThat(thrown).isInstanceOf(ModFileValidationError.class);

            var errorsCount = expectedErrors.size();

            var formattedErrors = expectedErrors.stream()
                    .map(error -> String.format(
                            "validation error at line=%d, column=%d: %s",
                            error.getLine().getStart(), error.getColumn().getStart(), error.getMessage()))
                    .collect(joining("\n\t* "));

            var expectedMessage = String.format(
                    "%d error%s occurred:\n\t* %s\n\n", errorsCount, errorsCount > 1 ? "s" : "", formattedErrors);

            assertThat(thrown).hasMessage(expectedMessage);
        }
    }

    private static Stream<Arguments> fgaModTestCases() {
        return TestsData.FGA_MOD_TRANSFORM_TEST_CASES.stream()
                .map(testCase -> arguments(
                        testCase.getName(),
                        testCase.getModFile(),
                        testCase.getJson(),
                        testCase.isSkip(),
                        testCase.getExpectedErrors()));
    }
}

File: pkg/csharp/test/OpenFga.Language.Tests/FgaModToJsonShould.cs
using Microsoft.VisualStudio.TestPlatform.CommunicationUtilities;
using OpenFga.Language.Errors;
using OpenFga.Language.ModFile;
using OpenFga.Language.Tests.util;
using OpenFga.Language.Transformers;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace OpenFga.Language.Tests;

public class FgaModToJsonShould
{
    [Theory]
    [MemberData(nameof(FgaModTestCases))]
    public void TransformFgaMod(
        string name, 
        string fgaMod, 
        string? json, 
        bool skip, 
        List<ModFileValidationSingleError> expectedErrors)
    {
        // Skip test if marked as skip
        if (skip)
        {
            return;
        }

        if (json != null)
        {
            var generatedJson = FgaModTransformer.Transform(fgaMod);
            var expected = Json.Stringify(Json.Parse<FgaModFile>(json));

            var different = expected != generatedJson;
            Assert.Equal(expected, generatedJson);
        }
        else
        {
            var exception = Assert.Throws<ModFileValidationError>(() => FgaModTransformer.Transform(fgaMod));

            var errorsCount = expectedErrors.Count;
            var formattedErrors = expectedErrors
                .Select(error => $"validation error at line={error.Line.Start}, column={error.Column.Start}: {error.Message}")
                .ToList();

            var expectedMessage = $"{errorsCount} error{(errorsCount > 1 ? "s" : "")} occurred:\n\t* {string.Join("\n\t* ", formattedErrors)}\n\n";

            var dif = expectedMessage != exception.Message;
            Assert.Equal(expectedMessage, exception.Message);
        }
    }

    public static IEnumerable<object[]> FgaModTestCases()
    {
        return TestsData.FgaModTransformTestCases
            .Select(testCase => new object[]
            {
                testCase.Name,
                testCase.ModFile,
                testCase.Json,
                testCase.Skip,
                testCase.ExpectedErrors
            });
    }
}
